### PR TITLE
doc: add compressdev programmers guide

### PR DESCRIPTION
**Pull request by ashishgupta4**
Earlier PR https://github.com/pablodelara/dpdk-draft-compressdev/pull/26 was closed by mistake. Submitting a new PR for the same. 

Add programmer’s guide for compress dev library.

Here's an updated programmer guides with
- review inputs
- description for hash support in compression API

Signed-off-by: Shally Verma <shally.verma@caviumnetworks.com>
Signed-off-by: Sunila Sahu <sunila.sahu@caviumnetworks.com>
Signed-off-by: Ashish Gupta <ashish.gupta@caviumnetworks.com>

### CODE REVIEW

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 1]
```diff
@@ -0,0 +1,1047 @@
...
+
+.. code-block:: c
+
+   int __rte_experimental  rte_compressdev_private_xform_create(uint8_t dev_id,
+                                        const struct rte_comp_xform *xform,
+                                        void **private_xform);
+
+   int __rte_experimental  rte_compressdev_private_xform_free(uint8_t dev_id, void *private_xform);
+
+An example sudo sample code to setup and process NUM_OPS stateless ops with each of length OP_LEN
```
Comment (pablodelara) [inline doc/guides/prog_guide/compressdev_lib.rst]: Pseudocode example?

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 2]
```diff
@@ -0,0 +1,1047 @@
...
+using shareable priv_xform would look like:
+
+.. code-block:: c
+
+	/*
+     * sudo example to do stateless compression
```
Comment (pablodelara) [inline doc/guides/prog_guide/compressdev_lib.rst]: Pseudo code example?

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 3]
```diff
@@ -0,0 +1,1047 @@
...
+        .next = NULL,
+		.type = RTE_COMP_COMPRESS,
+		.compress = {
+			.algo = RTE_COMP_ALGO_DEFLATE,
+			.deflate = {
+				.huffman = RTE_COMP_HUFFMAN_DEFAULT
+			},
+			.level = RTE_COMP_LEVEL_PMD_DEFAULT,
+			.chksum = RTE_COMP_CHECKSUM_NONE,
+			.window_size = DEFAULT_WINDOW_SIZE
```
Comment (pablodelara) [inline doc/guides/prog_guide/compressdev_lib.rst]: Add also hash field (probably to NONE, for simplicity)

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 4]
```diff
@@ -0,0 +1,1047 @@
...
+		non-shareable = 1;
+
+	/* Get a burst of operations. */
+    struct rte_comp_op *comp_ops[CHUNK_LEN];
+    if (rte_comp_op_bulk_alloc(op_pool, comp_ops, OP_LEN) == 0)
+        rte_exit(EXIT_FAILURE, "Not enough compression operations available\n");
+
+    /* Get a burst of src and dst mbufs. */
+
+    /* prepare source and destination mbufs for compression operations */
```
Comment (pablodelara) [inline doc/guides/prog_guide/compressdev_lib.rst]: Use capital letters for comments, for consistency.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 5]
```diff
@@ -0,0 +1,1047 @@
...
+    }
+
+    /* Set up the compress operations. */
+    for (i = 0; i < NUM_OPS; i++) {
+        struct rte_comp_op *op = comp_ops[i];
+		if (non-shareable)
+			rte_priv_xform_create(cdev_id, &compress_xform, &op->priv_xform)
+		else
+			op->priv_xform = priv_xform;
+		op->stream = stream;
```
Comment (pablodelara) [inline doc/guides/prog_guide/compressdev_lib.rst]: This example is for stateless operations, so stream is not required, but priv_xform is.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 6]
```diff
@@ -0,0 +1,1047 @@
...
+		op->m_src = src_buf[i];
```
Comment (pablodelara) [inline doc/guides/prog_guide/compressdev_lib.rst]: This was already done above.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 7]
```diff
@@ -0,0 +1,1047 @@
...
+		op->m_dst = dst_buf[i];
+		op->type = RTE_COMP_OP_STATELESS;
+		op->flush = RTE_COMP_FLUSH_FINAL;
+
+		op->src.offset = 0;
+		op->dst.offset = 0;
+		op->src.length = OP_LEN;
+		op->input_chksum = 0;
+	}
+	num_enqd = rte_compressdev_enqueue_burst(cdev_id, 0, &op[i],1);
```
Comment (pablodelara) [inline doc/guides/prog_guide/compressdev_lib.rst]: rte_compressdev_enqueue_burst(cdev_id, 0, comp_ops, NUM_OPS)

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 8]
```diff
@@ -0,0 +1,1047 @@
...
+	/* wait for this to complete before enqueing next*/
+	do {
+	num_deque = rte_compressdev_dequeue_burst(cdev_id, 0 , &processed_ops, 1);
```
Comment (pablodelara) [inline doc/guides/prog_guide/compressdev_lib.rst]: num_deqd = .... (..., NUM_OPS)

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 9]
```diff
@@ -0,0 +1,1047 @@
...
+	}while (num_dqud < num_enqd);
+
+
+Stateless and OUT_OF_SPACE
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+OUT_OF_SPACE is a condition when output buffer runs out of space and where PMD
+still has more data to produce. If PMD run into such condition, then it's an
```
Comment (pablodelara) [inline doc/guides/prog_guide/compressdev_lib.rst]: runs

Comment (pablodelara) [reply doc/guides/prog_guide/compressdev_lib.rst]: Then PMD returns RTE_COMP_OP...TERMINATED error.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 10]
```diff
@@ -0,0 +1,1047 @@
...
+error condition if PMD returns RTE_COMP_OP_OUT_OF_SPACE_TERMINATED.
+In such case, PMD resets itself and can set consumed=0 and produced=amount of output
+it could produce before hitting out_of_space. Application would need to
+resubmit an full input with larger output buffer size, if it want to operation
```
Comment (pablodelara) [inline doc/guides/prog_guide/compressdev_lib.rst]: resubmit the whole input with a larger output buffer, if it wants the operation to be completed.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 11]
```diff
@@ -0,0 +1,1047 @@
...
+.. code-block:: c
+
+   struct rte_mempool *rte_comp_op_pool_create(const char *name,
+						unsigned int nb_elts, unsigned int cache_size,
+						uint16_t user_size, int socket_id);
+
+
+``rte_comp_op_alloc()`` and ``rte_comp_op_bulk_alloc()`` are used to allocate
+compression operations from a given compression operation mempool.
+``__rte_comp_op_reset()`` is called on each operation before being returned to
```
[the remark this answers is not in the document]
Comment (ashishgupta4) [reply doc/guides/prog_guide/compressdev_lib.rst]: Actually, operation gets reset after it is allocated in rte_comp_op_alloc() and not when it is freed. 
I'll update this to "The operation gets reset before being returned to a user so that operation
is always in a good known state before use by the application."

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 12]
```diff
@@ -0,0 +1,1047 @@
...
+	out of space condition hit.	PMD has ability to recover
+	from here, so application can submit next op from
+	consumed = consumed+1 and an destination buffer with available space.
+
+Transforms and Transform Chaining
+----------------------------------
+
+Compression transforms (``rte_comp_xform``) are the mechanism
+to specify the details of the compression operation.
+Currently chaining is not supported on compression API.
```
[the remark this answers is not in the document]
Comment (ashishgupta4) [reply doc/guides/prog_guide/compressdev_lib.rst]: @pablodelara  I think xform chaining is not needed for compressdev and next pointer can be removed in rte_comp_xform.

File: doc/guides/prog_guide/compressdev_lib.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (pablodelara) [reply doc/guides/prog_guide/compressdev_lib.rst]: You are right, thanks.

File: doc/guides/prog_guide/compressdev_lib.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (pablodelara) [reply doc/guides/prog_guide/compressdev_lib.rst]: @1234sv Shally, do you agree to remove this? You had a use case in the past, but not sure if it is still the case. We could always add it later if needed.

[the remark this answers is not in the document]
Comment (1234sv) [reply doc/guides/prog_guide/compressdev_lib.rst]: I wanted to retain it for extensibility. However, agree I dont have immediate identified requirement for same. Would like to have @ahmedmpr input on same. Also, if it is  to be removed it then suggest raise PR with API change , if approved then update doc.

**Comment by ashishgupta4** [comment]
@pablodelara : Thanks for the review comments. I have incorporated these review and updated the docs. Please let me know if any other feedback there.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 13]
```diff
@@ -0,0 +1,1047 @@
...
+            shareable = 0;
+         }
+
+    if (mode == RTE_COMP_PRIV_XFORM_SHAREABLE)
+       rte_log(INFO, "using shareable priv_xform\n");
+    else {
+            rte_log(INFO, "priv_xform is non-shareable");
+            shareable = 0;
+         }
+
```
[the remark this answers is not in the document]
Comment (ashishgupta4) [reply doc/guides/prog_guide/compressdev_lib.rst]: Incorporated this review.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 14]
```diff
@@ -0,0 +1,1051 @@
...
+.. code-block:: c
+
+	struct rte_compressdev_info {
+		const char *driver_name;		/**< Driver name. */
+		uint64_t feature_flags;			/**< Feature flags */
+		const struct rte_compressdev_capabilities *capabilities;
+		/**< Array of devices supported capabilities */
+		uint16_t max_nb_queue_pairs;
+		/**< Maximum number of queues pairs supported by device. */
+		/**< Maximum number of queues pairs supported by device.
```
Comment (1234sv) [inline doc/guides/prog_guide/compressdev_lib.rst]: Comment got repeated here

File: doc/guides/prog_guide/compressdev_lib.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (ahmedmpr) [reply doc/guides/prog_guide/compressdev_lib.rst]: That is a cool feature, but no immediate use case that I can think of besides debugging. Ok with removing it

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 15]
```diff
@@ -0,0 +1,1051 @@
...
+to perform those services. The framework currently only supports lossless compression schemes:
+Deflate and LZS.
+
+Device Management
+-----------------
+
+Device Creation
+~~~~~~~~~~~~~~~
+
+Physical compression devices are discovered during the PCI probe/enumeration of the
```
[the remark this answers is not in the document]
Comment (ahmedmpr) [reply doc/guides/prog_guide/compressdev_lib.rst]: agreed, we do not use PCI

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 16]
```diff
@@ -0,0 +1,1051 @@
...
+.. code-block:: console
+
+   --vdev  '<pmd name>,socket_id=0'
+
+.. Note::
+
+   * If DPDK application requires multiple software compression PMD devices then required
+     number of ``--vdev`` with appropriate libraries are to be added.
+
+   * An Application with compression PMD instaces sharing the same library requires unique ID.
```
[the remark this answers is not in the document]
Comment (1234sv) [reply doc/guides/prog_guide/compressdev_lib.rst]: No that's not what I suggest. I agree that PMDs should register with unique name. However I am bit confused by the reference of "same library" in sentence here. Say, if there're two PMDs both using ZLIB underneath, they should use unique names but those could be anything, such as, "compress_zlib0" "compress_def" or  "compress_zlib0" "compress_zlib1". So, just want to make sure, there are no other guidelines on naming except that they should be unique.

[the remark this answers is not in the document]
Comment (fionatrahe) [reply doc/guides/prog_guide/compressdev_lib.rst]: ah, ok. I guess the comment is in the context of the -vdev params. So if only 1 device from each PMD is specified, then there's no need to specify a device name as the PMD name will be used. But if more than one device from the same PMD is specified, then they need to specify separate device names on the vdev command param. Either way the result is a unique name for each device.
So maybe reword as 
An Application with multiple compression device instances exposed by the same PMD must specify a unique name for each device.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 17]
```diff
@@ -0,0 +1,1051 @@
...
+Logical Cores, Memory and Queues Pair Relationships
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Library supports NUMA similarly as described in Cryptodev library section.
+
+Multiple logical cores should never share the same queue pair for enqueuing
+operations or dequeuing operations on the same compression device since this would
+require global locks and hinder performance. It is however possible to use a
+different logical core to dequeue an operation on a queue pair from the logical
+core which it was enqueued on. This means that a compression burst enqueue/dequeue
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: grammar: on which it was enqueued. :)

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 18]
```diff
@@ -0,0 +1,1051 @@
...
+The ``max_inflight_ops`` is used to pass maximum number of
+rte_comp_op that could be present in a queue at-a-time.
+PMD then can allocate resources accordingly on a specified socket.
+
+Logical Cores, Memory and Queues Pair Relationships
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Library supports NUMA similarly as described in Cryptodev library section.
+
+Multiple logical cores should never share the same queue pair for enqueuing
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: This is true not just of logical cores, but also of processes or threads.  A queue pair cannot be shared and should be exclusively used by a single processing context. Except for the exception you've listed.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 19]
```diff
@@ -0,0 +1,1051 @@
...
+APIs are a logical place to transition from one logical core to another in a
+data processing pipeline.
+
+Device Features and Capabilities
+---------------------------------
+
+Compression devices define their functionality through two mechanisms, global device
+features and algorithm features. Global devices features identify device
+wide level features which are applicable to the whole device such as supported hardware
+acceleration and compression algorithms.
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: not "and compression algorithms"
Maybe "and CPU features"

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 20]
```diff
@@ -0,0 +1,1051 @@
...
+
+The algorithm features lists individual algo feature which device supports,
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: Maybe ...lists individual features which the device supports per-algorithm, such as.....

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 21]
```diff
@@ -0,0 +1,1051 @@
...
+such as a stateful compression/decompression, checksums operation etc.
+
+Device Features
+~~~~~~~~~~~~~~~
+
+Following are current Compression device feature flags:
+
+* SSE accelerated SIMD vector operations
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: I think it's better not to list this level of detail here, it'll easily get out of sync with code. The examples you listed above are enough. Maybe mention instead that the first can be seen in the list of RTE_COMPDEV_FF_XXX and the latter in the RTE_COMP_FF_XXX list

File: doc/guides/prog_guide/compressdev_lib.rst (hunk 19 above)
Comment (1234sv) [reply doc/guides/prog_guide/compressdev_lib.rst]: ok

File: doc/guides/prog_guide/compressdev_lib.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (pablodelara) [reply doc/guides/prog_guide/compressdev_lib.rst]: I created a PR for this, so I will merge it then.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 22]
```diff
@@ -0,0 +1,1051 @@
...
+	};
+
+Compression Operation
+----------------------
+
+DPDK compression supports two types of compression methodologies:
+
+- Stateless, each data object is compressed individually without any reference to previous data
+
+- Stateful, each data object is compressed with reference to previous data object i.e. history of data is needed for compression / decompression.
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: This is not quite correct. e..g. in SGL - aka chained mbuf  - case, the objects can be chained together and submitted as a STATELESS operation. The STATEFUL attribute in compressdev API refers to data being stateful across a sequence of comp_ops.  SO maybe:
data in each operation is compressed with reference to previous operations in the same data stream i.e. history of data is maintained between operations.

I'd leave out the mention of compression / decompression to avoid any confusion of stateful compression implying stateful decompression.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 23]
```diff
@@ -0,0 +1,1051 @@
...
+
+For more explanation, please refer RFC https://www.ietf.org/rfc/rfc1951.txt
+
+Operation Representation
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Compression operation is described via ``struct rte_comp_op``. The operation structure
+includes the operation type (stateless or stateful), the operation status
+and the priv_xform/stream handle, source, destination and checksum buffer
+pointers. It also contains the source mempool for the operation are allocated
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: grammar: ...mempool from which the operation is allocated.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 24]
```diff
@@ -0,0 +1,1051 @@
...
+
+- Stateful, each data object is compressed with reference to previous data object i.e. history of data is needed for compression / decompression.
+
+For more explanation, please refer RFC https://www.ietf.org/rfc/rfc1951.txt
+
+Operation Representation
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Compression operation is described via ``struct rte_comp_op``. The operation structure
+includes the operation type (stateless or stateful), the operation status
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: Maybe worth adding: .....rte_comp_op, which contains both input and output data.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 25]
```diff
@@ -0,0 +1,1051 @@
...
+and the priv_xform/stream handle, source, destination and checksum buffer
+pointers. It also contains the source mempool for the operation are allocated
+from. PMD consumes the input as mentioned in consumed field and update
+produced with amount of data of written into destination buffer along with
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: This makes it sound like "consumed" value is input and only produced is updated. Both are output params which are updated.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 26]
```diff
@@ -0,0 +1,1051 @@
...
+status of operation.
+
+Compression operations mempool also has an ability to allocate private memory with the
+operation for applications purposes. Application software is responsible for specifying
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: grammar: application's

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 27]
```diff
@@ -0,0 +1,1051 @@
...
+all the operation specific fields in the ``rte_comp_op`` structure which are then used
+by the compression PMD to process the requested operation.
+
+.. code-block:: c
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: I don't think it's a good idea to copy this level of code detail - better to refer to the code. Else it'll get out of sync in future.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 28]
```diff
@@ -0,0 +1,1051 @@
...
+
+- If status is RTE_COMP_OP_STATUS_SUCCESS,
+	consumed = amount of data read from input buffer, and
+	produced = amount of data written in destination buffer
+- If status is RTE_COMP_OP_STATUS_FAILURE,
+	consumed = produced = 0 or undefined
+- If status is RTE_COMP_OP_STATUS_OUT_OF_SPACE_TERMINATED,
+	consumed = 0 and
+	produced = amount of data successfully produced until
+	out of space condition hit.	Application can consume output data, if required.
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: would add typically produced =0, not to give the impression there is usually output data.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 29]
```diff
@@ -0,0 +1,1051 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+OUT_OF_SPACE is a condition when output buffer runs out of space and where PMD
+still has more data to produce. If PMD runs into such condition, then PMD returns
+RTE_COMP_OP_OUT_OF_SPACE_TERMINATED error. In such case, PMD resets itself and can set
+consumed=0 and produced=amount of output it could produce before hitting out_of_space.
+Application would need to resubmit the whole input with a larger output buffer, if it
+wants the operation to be completed.
+
+Hash in Stateless
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: Mention that same applies to checksum.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 30]
```diff
@@ -0,0 +1,1051 @@
...
+In such case
+- ops are setup with op_type RTE_COMP_OP_STATEFUL,
+- all ops except last set to flush value = RTE_COMP_NO/SYNC_FLUSH
+and last set to flush value RTE_COMP_FULL/FINAL_FLUSH.
+
+In case of either one or all of the above conditions, PMD initiates
+stateful processing and releases acquired resources after processing
+operation with flush value = RTE_COMP_FLUSH_FULL/FINAL is complete.
+Unlike stateless, application can enqueue only one stateful op from
+a particular stream in a single burst and must attach stream handle
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: better to say ...application can enqueue only one stateful op from a particular stream at a time and must attach... 
i.e. whether it's in a single burst or not is not the point. The next op can't be sent in the next burst unless the last op is dequeued,

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 31]
```diff
@@ -0,0 +1,1051 @@
...
+
+Unlike priv_xforms, stream is always a NON_SHAREABLE entity. One stream handle must be attached to only
+one set of related ops and cannot be reused until all of them are processed with status Success or failure.
+
+.. figure:: img/stateful-op.png
+
+Application should call ``rte_comp_stream_create()`` and attach to op before
+enqueuing them for processing and free via ``rte_comp_stream_free()`` during
+termination. All ops that are to be processed statefully should carry *same* stream.
+
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: Maybe add a note somewhere : A burst may contain multiple stateful ops as long as each op is attached to a different stream.

File: doc/guides/prog_guide/compressdev_lib.rst [hunk 32]
```diff
@@ -0,0 +1,1051 @@
...
+A burst in DPDK compression can be a combination of stateless and stateful operations with a condition
+that for stateful ops only one op at-a-time should be enqueued from a particular stream i.e. no-two ops
+should belong to same stream in a single burst i.e. a burst can look like:
+
++--------------+-------------+--------------+-----------------+--------------+--------------+
+|enqueue_burst |op1.no_flush | op2.no_flush | op3.flush_final | op4.no_flush | op5.no_flush |
++--------------+-------------+--------------+-----------------+---------------+-------------+
+
+Where, op1 .. op5 all belong to different independent data units and can be of type : stateless or stateful.
```
Comment (fionatrahe) [inline doc/guides/prog_guide/compressdev_lib.rst]: op1,2,4,5 must be stateful in this example as stateless ops can only use flush full or final.

**Comment by 1234sv** [comment]
Thanks for reviewing. I agree to your point to remove structs defines ( rte_comp_op / rte_comp_xforms et el ) from here and assume same is applicable to various API prototypes listed in docs? Will try to simplify pseudo codes illustrating stream and private_xforms usage.

**Comment by ashishgupta4** [comment]
@fionatrahe Thanks for the review. I have incorporated these changes and updated the docs.

File: doc/guides/prog_guide/compressdev_lib.rst (hunk 29 above)
Comment (ashishgupta4) [reply doc/guides/prog_guide/compressdev_lib.rst]: Added a section below for the same.

File: doc/guides/prog_guide/compressdev_lib.rst (hunk 31 above)
Comment (ashishgupta4) [reply doc/guides/prog_guide/compressdev_lib.rst]: Added a note regarding this in section 'Burst in compression API'

**Comment by ashishgupta4** [comment]
Converted the PNG images to SVG format.

**Comment by 1234sv** [comment]
How did you check for acceptability? is there any script that we can run and confirm before sending further update?

**Comment by pablodelara** [comment]
As far as I know, there is no script to check the format.
Looking at http://dpdk.org/doc/guides/contributing/documentation.html, section 4.5.6 (Images):

All images should be in SVG scalar graphics format. They should be true SVG XML files and should not include binary formats embedded in a SVG wrapper.

Also, before sending another update, check that the docs built OK, with "make doc-guides-html".

**Comment by 1234sv** [comment]
Ok. Sorry to miss this info. Will take care going further.

**Comment by ashishgupta4** [comment]
@pablodelara I've updated the docs with SVG images. Please let me know if these are fine.

**Comment by pablodelara** [comment]
Thanks @ashishgupta4. The format of the SVG files look good now, thanks!
Minor comments:
- In the "shared stateless ops" figure, priv_xform could be together in one line. The "m" is in a second line unnecessarily, I think
- In the "non-shared stateless ops" figure, there is a typo in "priv_xfrm"
- In the "stateful ops" figure, I think the word "stream" on the left is not required, looks duplication.

- In "Device Configuration" section, rte_compressdev_config also has maximum number of private xforms and streams
- In "Burst in compression API" section, in the table showing an array of operations, leave a space on both sides of each word.

Lastly, I was looking at the pseudo code in the documentation, and reviewing the xform creation, I realized that ".hash_algo = RTE_COMP_HASH_ALGO_UNSPECIFIED" looks odd. Since it should mean that no hashing will be done, I think we could change that to "hash_algo = RTE_COMP_HASH_ALGO_NONE".
What do you think?

Thanks!

**Comment by 1234sv** [comment]
Thanks for feedback. Few comments:
`In` "Device Configuration" section, rte_compressdev_config also has maximum number of private xforms and `streams`
-- mentioning max num streams/private xforms here will make doc specific to structure definition. So, I deliberately avoiding referring to that and instead put generic information , such as,  PMD allocate its resources  and mention reference to rte_compressdev_configure structure instead. 
For hash_algo = RTE_COMP_HASH_ALGO_UNSPECIFIED, currently API patch 8/13, doesn't have enum name HASH_NONE. If , change is desired, then we will change the API first and then doc. Till then, lets merge it as is to be in sync with Patch v5. In any case, change is trivial and shouldn't be a blocker for this doc to merge.
Will take care of rest of the inputs. Assume, these are last set of changes proposed and we could be able to close on this once newer changes are approved.

**Comment by pablodelara** [comment]
Agree, definitely not a blocker for this doc. Actually, once these changes are made, let's merge this and send a v6. If there are more comments, we can make them in the mailing list.
About the hash macro, we could make the change in both the API and doc.
About the device configuration,  it just looks like it only has those two parameters. If you want to have generic comments, then probably you shouldn't call out two of the parameters and not the others. Maybe better to not call out any of them then.

Thanks for the work!

**Comment by 1234sv** [comment]
Am bit confused to understand 
>then probably you shouldn't call out two of the parameters and not the others. Maybe better to not call out any of them then.
Where are we calling config specific parameters ? could you point in doc where this comment is applicable to help me understand better?
We will update PR with changes on figures. However not planning to do Hash changes as of now. Can take them separately after 1st cut sometime.

**Comment by pablodelara** [comment]
When you say "The ``rte_compressdev_config`` structure is used to pass the configuration
parameters for socket selection and number of queue pairs." 
About Hash, that's ok for me. Actually, if you agree on changing UNSPECIFIED to NONE, I can make the change on the fly in v6, and stop polluting this with more PRs :)

**Comment by 1234sv** [comment]
Oh..ok got it. Will rephrase it. for hash, yes that sounds like a better plan.

**Comment by ashishgupta4** [comment]
@pablodelara Following are the updates:
- svg files are updated based on the suggestions.
- Modified device configuration section
- Ops table is updated in "Burst in compression API"  section.
- hash enum can be changed from UNSPECIFIED to NONE with API changes.
